src/iop/liquify.c

```diff
@@ -2268,9 +2183,219 @@ static float find_nearest_on_curve_t(const float complex p0,
+  const float b     = cabsf(p1 - p0);         // |b|
+  const float dotab = cdot(x - p0, p1 - p0);  // |a| * |b| * cos(phi)
+  return dotab / (b * b);                     // |a| / |b| * cos(phi)
+}
+
+// split a cubic bezier at t into two cubic beziers.
+
+static void casteljau(const float complex *p0, float complex *p1, float complex *p2, float complex *p3, const float t)
+{
+  const float complex p01 = *p0 + (*p1 - *p0) * t;
+  const float complex p12 = *p1 + (*p2 - *p1) * t;
+  const float complex p23 = *p2 + (*p3 - *p2) * t;
+
+  const float complex p012 = p01 + (p12 - p01) * t;
+  const float complex p123 = p12 + (p23 - p12) * t;
+
+  const float complex p0123 = p012 + (p123 - p012) * t;
+
+  *p1 = p01;
+  *p2 = p012;
+  *p3 = p0123;
+}
+
+#define CHECK_HIT_PT(point)             \
+  const float d = cabsf(point - (*pt)); \
+  if(d < distance)                      \
+  {                                     \
+    distance = d;                       \
+    hit.layer = layer;                  \
+    hit.elem = data;                    \
+  }
+
+static dt_liquify_hit_t _hit_paths(dt_iop_module_t *module,
+                                   dt_iop_liquify_params_t *p,
+                                   GList *layers,
+                                   const float complex *pt)
+{
+  dt_liquify_hit_t hit = NOWHERE;
+
+  float distance = FLT_MAX;
+
+  for(GList *l = layers; l != NULL; l = l->next)
+  {
+    const dt_liquify_layer_enum_t layer = (dt_liquify_layer_enum_t) GPOINTER_TO_INT(l->data);
+
+    if((dt_liquify_layers[layer].flags & DT_LIQUIFY_LAYER_FLAG_HIT_TEST) == 0)
+      continue;
+
+    for(int k=0; k<MAX_NODES; k++)
+    {
+      dt_liquify_path_data_t *data = &p->nodes[k];
+      const dt_liquify_path_data_t *prev = node_prev(p, data);
+
+      if(p->nodes[k].header.type == DT_LIQUIFY_PATH_INVALIDATED)
+        break;
+
+      if((dt_liquify_layers[layer].flags & DT_LIQUIFY_LAYER_FLAG_NODE_SELECTED)
+          && !data->header.selected)
+        continue;
+
+      if((dt_liquify_layers[layer].flags & DT_LIQUIFY_LAYER_FLAG_PREV_SELECTED)
+          && (!prev || !prev->header.selected))
+        continue;
+
+      const dt_liquify_warp_t *warp  = &data->warp;
+      const float complex point = data->warp.point;
+
+      if(layer == DT_LIQUIFY_LAYER_PATH)
+      {
+        if(data->header.type == DT_LIQUIFY_PATH_LINE_TO_V1)
+        {
+          // remove 5% from start and end of line as non sensible area
+          // this is to avoid wrong interaction for center point on both
+          // sides.
+          const float complex deadzone = (point - prev->warp.point) / 20.0f;
+          const float complex lp1 = prev->warp.point + deadzone;
+          const float complex lp2 = point - deadzone;
+          const float t = find_nearest_on_line_t(lp1, lp2, *pt);
+
+          if(t > 0.0f && t < 1.0f)
+          {
+            const float complex linepoint = cmix(lp1, lp2, t);
+            const float d = cabsf(linepoint - *pt);
```

Comment:
d should be a euclidean distance, no ? aka hypotf(x_cursor - x_node, y_cursor - y_node).

Reply:
if I'm not mistaken this is what cabsf() does here.

Reply:
The `cabs(z)` function is equivalent to `hypot(z.x, z.y)`, so you're both right :smile:

Reply:
I'm just concerned that dealing with complex numbers will be slower than dealing with a vector of floats

Reply:
No, behind the scene this is just a set of two float. Nothing more, nothing less and all nice routines coming with them are very useful in a two space canvas.